### PR TITLE
Fix server startup on NetBSD (and probably other systems)

### PR DESCRIPTION
**Pull request by manu0401**
my_open_parent_dir_nosymlinks() calls openat() with dfd = -1, which
Linux takes for root of filesystem, but this behavior is not standard
as per OpenGroup XSH:
http://pubs.opengroup.org/onlinepubs/9699919799/functions/open.html

On NetBSD, dfd = -1 causes openat() to fail with EBADF, which breaks
server startup. The same result probaly happens on any non Linux
system implementing openat().

The proposed fix is to call plain open() when dfd = -1

### CODE REVIEW

File: mysys/my_symlink.c [hunk 1]
```diff
@@ -244,7 +244,20 @@ const char *my_open_parent_dir_nosymlinks(const char *pathname, int *pdfd)
...
+     * http://pubs.opengroup.org/onlinepubs/9699919799/functions/open.html
+     * At least on NetBSD, calling openat() with dfd = -1 fails with EBADF.
+     * Since s has been tested to be an absolute parth, just use open()
+     * on non Linux systems for the same result.
+     */
+#ifndef linux
+    if (dfd == -1)
+      fd = open(s, O_NOFOLLOW | O_PATH | O_CLOEXEC);
+    else
+#endif
```
Comment (svoj) [inline mysys/my_symlink.c]: I'd suggest to move it before the loop and instead do something like:
```
#ifndef
dfd= open("/", O_NOFOLLOW | O_PATH | O_CLOEXEC);
#endif
```

**Comment by vuvova** [comment]
Strictly speaking, Linux does not take `dfd==-1` as root. The manual page `open(2)` says:
> If pathname is absolute, then dirfd is ignored.

The OpenGroup page you mentioned says
> The openat() function shall be equivalent to the open() function except in the case where path specifies a relative path.

As far as I understand, it means that if the path is "/", dfd does not matter and could as well be -1.
If NetBSD behaves differently, it must be a NetBSD bug.

**Comment by manu0401** [comment]
On Tue, Jan 29, 2019 at 09:18:30AM +0000, Sergei Golubchik wrote:
> If NetBSD behaves differently, it must be a NetBSD bug.

Well, then we will fix NetBSD, but if you want to support 
relased NetBSD so far, the workaround is required.

-- 
Emmanuel Dreyfus
manu@netbsd.org

**Comment by vuvova** [comment]
Yes, I tend to agree with that

**Comment by LinuxJedi** [comment]
Hello @manu0401,

I'm trying to go through older pull requests so that we can work towards getting them moving again. Is this patch still something you would like to see in MariaDB? If so I can try to assist you in getting it to a place where it can be merged.

Please let me know how you would like to proceed.

**Comment by LinuxJedi** [comment]
Hello @manu0401,

I'm going to close this one for now as we have not heard from you. But please feel free to reopen or contact me if you wish to continue this one or discuss it further.

**Pull request closed without merging**